### PR TITLE
#1004 Add new filter to remove elements in array

### LINKED ISSUE #1004: Filter out results on arrays before displaying on Plugin element
Required specifically for removing the `admin` preferences to be selected in templates e.g. GrantUserPermission.

### Existing implementation

Currently the list of permissions is displayed by `Checkbox` plugin element, which only uses an array passed on (or received by a API query result) to be displayed, there is no filters applied.
Screen Shot 2023-01-17 at 1 48 06 PM

The current API query used will return ALL permissions based on the orgId and/or userId received:
Screen Shot 2023-01-17 at 1 48 30 PM

It would be tricky to add some filtering to the `Checkbox` plugin or to change the existing `user-permissions` end-point

### Proposed Solution
Create a new evaluator custom function to filter an array, which would be what the `checkboxes` parameter would call, and pass as one of the parameters the result from the API query to `user-permissions`. 

Syntax:

```
filter( valuesArray, options }

Where "options" is:
{
  key?: string
  values: string | string[] (in principle could be numbers or any primitive type)
  rule?: "exclude" | "include (default: "exclude")
}
```

Parameters types explained:
`valuesArray`: array of objects OR strings
`key`: which key to filter (optional -- omitted means the input is an array of strings/primitives not array of objects
`values`: array of values to filter OR single value
`rule`: enum to say if should include or exclude these keys on the result (default "exclude" would you say?)

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #1004

Required a new evaluator function to remove permissions from the array returned by user-permissions, so not all of the permissions displays in specific templates. Work to be used in configuration of https://github.com/openmsupply/conforma-templates/issues/55

### CODE REVIEW

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [inline src/components/actions/evaluatorFunctions.ts:14]: Why not just:
```
rule?: 'exclude' | 'include'
```

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [inline src/components/actions/evaluatorFunctions.ts:21]: `valuesArray.length > 0`

No need for this restriction, is there?

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [inline src/components/actions/evaluatorFunctions.ts:18]: I don't think it should be restricted to just objects and strings -- any primitive type should work fine.

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [inline src/components/actions/evaluatorFunctions.ts:27]: `obj // Array of strings`
Don't you mean "Array of objects"?

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [inline src/components/actions/evaluatorFunctions.ts:38]: It's nice `values.includes` works for both arrays and strings, but unfortunately it doesn't work for other primitive types.

Probably best to convert it all to an array at the top:
```
const compareValues = Array.isArray(values) ? values : [ values ]
```

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [reply src/components/actions/evaluatorFunctions.ts:21]: Probably no need to check the type of the first item.

I would just check *each* item as part of the filter and if it's not an object then:
- exclude it if the rule is "include"
- include it if the rule is "exclude"

Then it follows the rule in a way that makes sense regardless of what type it is :)

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [inline src/components/actions/evaluatorFunctions.ts:42]: Following the type check suggestion above, you won't need this at all.

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [inline src/components/actions/evaluatorFunctions.ts:25]: Also, with proper type checking within the filter, you won't need the `as`.

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Cool, thanks. Seems to mostly be working okay. Few suggestions added.

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [reply src/components/actions/evaluatorFunctions.ts:18]: This is a good case for using `valuesArray: unknown[]` and then having proper type guards within (as suggested above).

**Comment by nmadruga** [comment]
Carl I tried to address your suggestions in this commit [aa57ea1](https://github.com/openmsupply/conforma-server/pull/1005/commits/aa57ea142fb1b0500ba2c8a0fd01bde38af15e23)

**Comment by CarlosNZ** [comment]
> Carl I tried to address your suggestions in this commit [aa57ea1](https://github.com/openmsupply/conforma-server/pull/1005/commits/aa57ea142fb1b0500ba2c8a0fd01bde38af15e23)

Cool, looking really nice. Looks like it's all correct, but would you mind copying to the front-end branch as well so I can just do some quick tests in the UI?

File: src/components/actions/evaluatorFunctions.ts
Comment (CarlosNZ) [inline src/components/actions/evaluatorFunctions.ts:20]: `typeof element === 'object'` isn't *quite* enough as both `null` and arrays return true.

It'll need to be `typeof element === 'object' && !Array.isArray(element) && element !== null`

Feel free to abstract this out into a `isObject` utility function if you like.

Comment (nmadruga) [reply src/components/actions/evaluatorFunctions.ts:20]: Hmmm no problem. I'll do this now

**Review by CarlosNZ — APPROVED** [review]
Cool, thanks. 

See my tiny comment for front-end.

Approved, assuming both front and back-end will be made identical before merging :) 

👍

**Pull request merged**
